File: leetcode/0523-continuous-subarray-sum/solution.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool checkSubarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> firstIndex;
        firstIndex[0] = -1;

        long long prefixSum = 0;

        for (int i = 0; i < (int)nums.size(); i++) {
            prefixSum += nums[i];
            int remainder = (int)(prefixSum % k);

            auto it = firstIndex.find(remainder);
            if (it != firstIndex.end()) {
                if (i - it->second >= 2) {
                    return true;
                }
            } else {
                firstIndex[remainder] = i;
            }
        }

        return false;
    }
};

File: leetcode/0523-continuous-subarray-sum/solution.py
from typing import List


class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        # first index at which each prefix-sum-mod-k value was seen
        first_index = {0: -1}
        prefix_sum = 0

        for i, x in enumerate(nums):
            prefix_sum += x
            remainder = prefix_sum % k if k != 0 else prefix_sum

            if remainder in first_index:
                if i - first_index[remainder] >= 2:
                    return True
            else:
                first_index[remainder] = i

        return False
